keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2168 @@
+   application. This has been discussed in the issue [kubernetes/kubernetes#75482](https://github.com/kubernetes/kubernetes/issues/75482)
+   and in the issue [kubernetes-sigs/cluster-api#6158](https://github.com/kubernetes-sigs/cluster-api/issues/6158).
+4. There are cases during a node shutdown, when data corruption can occur due to premature node
+   shutdown. It would be great if applications could perform data migration and synchronization of 
+   cached writes to the underlying storage before the pod deletion occurs. This is not easy to
+   quantify even with pod's `.spec.shutdownGracePeriod`, as the time depends on the size of the data
+   and the speed of the storage. This has been discussed in the issue [kubernetes/kubernetes#116618](https://github.com/kubernetes/kubernetes/issues/116618)
+   and in the issue [kubernetes/kubernetes#115148](https://github.com/kubernetes/kubernetes/issues/115148).
+5. During the Graceful Node Shutdown the kubelet terminates the pods in order of their priority.
+   The DaemonSet controller runs its own scheduling logic and creates the pods again. This causes a
+   race. Such pods should be removed and not recreated, but higher priority pods that have not yet
+   been terminated should be recreated if they are missing. This has been discussed in the issue
+   [kubernetes/kubernetes#122912](https://github.com/kubernetes/kubernetes/issues/122912).
+6. The Graceful Node Shutdown feature is not always reliable. If Dbus or kubelet is restarted
+   during the shutdown, pods may be ungracefully terminated, leading to application disruption and
+   data loss. New applications can get scheduled on such a node which can also be harmful.
+   This has been discussed in issues [kubernetes/kubernetes#122674](https://github.com/kubernetes/kubernetes/issues/122674),
+   [kubernetes/kubernetes#120613](https://github.com/kubernetes/kubernetes/issues/120613) and [kubernetes/kubernetes#122674](https://github.com/kubernetes/kubernetes/issues/112443).
+7. There is no way to gracefully terminate static pods during a node shutdown
+   [kubernetes/kubernetes#122674](https://github.com/kubernetes/kubernetes/issues/122674), and the
+   lifecycle/termination is not clearly defined for static pods [kubernetes/kubernetes#16627](https://github.com/kubernetes/kubernetes/issues/16627).
+8. Different pod termination mechanisms are not synchronized with each other. So for example, the
+   taint manager may prematurely terminate pods that are currently under Node Graceful Shutdown.
+   This can also happen with other mechanism (e.g., different types of evictions). This has been
+   discussed in the issue [kubernetes/kubernetes#124448](https://github.com/kubernetes/kubernetes/issues/124448)
+   and in the issue [kubernetes/kubernetes#72129](https://github.com/kubernetes/kubernetes/issues/72129).
+9. There is not enough metadata about why the node drain was requested. This has been discussed in
```

Comment:
A lot of the above points nicely collected issues with Graceful Node Shutdown, which is being worked on as part of https://github.com/kubernetes/enhancements/issues/2000. Have you considered pointing the authors to this list?

Reply:
There has been a cooperation/discussion with sig-node. I have discussed this with @kwilczynski, who is interested in graduating the Graceful Node Shutdown feature. 

Many of the problems described are impossible or impractical to be solved by the kubelet alone. This is one of the main goals of this KEP to help resolve this situation.

cc @SergeyKanzhelev @dchen1107